### ISSUE TITLE
[NLP] punctuation_capitalization resume training fails

### ISSUE THREAD

**Issue by itzsimpl** [NVIDIA/NeMo#4522]
**Describe the bug**

When training PC from scratch, using the example configs, everything works fine, however if I try to resume an interrupted training from the last checkpoint, I get the following log entry
```
[NeMo W 2022-07-07 21:18:38 nemo_logging:349] /opt/conda/lib/python3.8/site-packages/pytorch_lightning/callbacks/model_checkpoint.py:378: UserWarning: `ModelCheckpoint(monitor='val_loss')` could not find the monitored key in the returned metrics: ['epoch', 'step']. HINT: Did you call `log('val_loss', value)` in the `LightningModule`?
      warning_cache.warn(m)
```
and all new checkpoints get a `val_loss=0.0000`

**Environment details**

- pytorch:22.04-py3
- nemo:1.8.2 installed with ./reinstall.sh, without numba install

**Comment by PeganovAnton** [comment]
Installed `nemo` as you described and could not reproduce the bug. Could you please describe exact setting?

**Comment by PeganovAnton** [comment]
@itzsimpl

**Comment by itzsimpl** [comment]
@PeganovAnton I apologise for taking so long, I was trying to debug. I haven't found the root cause, but by changing to pytorch:22.06-py3 and nemo:1.10.0, I don't see the issue anymore. I'll just continue with nemo:1.10.0. I did a couple of minor modifications to the model codebase (support for Uppercase, Lowercase and Capitalised), but these should not have any influence. Especially since I ported them to nemo:1.10.0 and everything works as it should.

**Comment by itzsimpl** [comment]
FWW. I was running the following command:
```
python examples/nlp/token_classification/punctuation_capitalization_train_evaluate.py \
  --config-path=conf \
  --config-name=punctuation_capitalization_config \
  \
  exp_manager.name=punctuation_capitalization_v4 \
  exp_manager.exp_dir=/experiments \
  +exp_manager.version=20220629-1159 \
  +exp_manager.resume_if_exists=true \
  +exp_manager.resume_ignore_no_checkpoint=true \
  +exp_manager.checkpoint_callback_params.save_best_model=true \
  \
  model.language_model.pretrained_model_name=bert-base-multilingual-cased \
  model.train_ds.ds_item=/data/punct_v4 \
  model.train_ds.text_file=text_train.txt \
  model.train_ds.labels_file=labels_train.txt \
  model.train_ds.use_tarred_dataset=True \
  model.train_ds.tar_metadata_file=tarred_train/metadata.punctuation_capitalization.tokens15000.max_seq_length512.bert-base-multilingual-cased.json \
  model.validation_ds.ds_item=/data/punct_v4 \
  model.validation_ds.text_file=text_val.txt \
  model.validation_ds.labels_file=labels_val.txt \
  model.validation_ds.use_tarred_dataset=True \
  model.validation_ds.tar_metadata_file=tarred_val/metadata.punctuation_capitalization.tokens15000.max_seq_length512.bert-base-multilingual-cased.json \
  ~model.test_ds \
  \
  trainer.devices=-1 \
  trainer.accelerator=gpu \
  trainer.num_nodes=1 \
  trainer.precision=bf16 \
  trainer.max_epochs=150 \
  trainer.max_steps=-1 \
  +trainer.num_sanity_val_steps=0 \
  \
  +exp_manager.checkpoint_callback_params.monitor=val_loss \
  +exp_manager.checkpoint_callback_params.mode=min \
  +exp_manager.checkpoint_callback_params.save_top_k=5 \
  +exp_manager.checkpoint_callback_params.always_save_nemo=true \
  +do_training=true \
  +do_testing=false \
  ```

**Comment by itzsimpl** [comment]
I'm really puzzled; if I set `trainer.val_check_interval` to e.g. 0.25 or anything less than 1.0 (as set in config yaml), I see the `Validation Dataloader` start and finish, and a valid checkpoint will be saved. 

If, however, I set it to 1.0, the `Validation Dataloader` will not start, the `val_loss` UserWarning will appear, and the checkpoint will be saved with `val_loss=0.0000`.

My changes to the codebase are not at fault, as they are a two line substitute for  
https://github.com/NVIDIA/NeMo/blob/dc6bea2acbfff899f06d3d7fa6ac98aec3aa4af3/nemo/collections/nlp/models/token_classification/punctuation_capitalization_model.py#L1005

Any thoughts @PeganovAnton?

**Comment by itzsimpl** [comment]
I've just noticed another strange behaviour, with `trainer.val_check_interval=0.25` the Epoch progress in the logs does not reach 100%; as if the last validation check is still not run, i.e. I see the first 3 validations of size 3125 (starting at 15%, at 39% and at 64%) in the logs, but not the last one.

```
Epoch 0:  88%|████████▊ | 27000/30750 [50:49<07:03,  8.85it/s, loss=0.117, v_num=1352, lr=5.26e-5]
Epoch 0:  88%|████████▊ | 27000/30750 [50:50<07:03,  8.85it/s, loss=0.117, v_num=1352, lr=5.26e-5]
Epoch 0:   0%|          | 0/30750 [00:00<?, ?it/s, loss=0.117, v_num=1352, lr=5.26e-5]            
Epoch 1:   0%|          | 0/30750 [00:00<?, ?it/s, loss=0.117, v_num=1352, lr=5.26e-5]
```

**Comment by itzsimpl** [comment]
@PeganovAnton Ok, noted. Thanks for the input. 

I started off with normal datasets, and moved to tarred ones based on the discussion in #4410 (https://github.com/NVIDIA/NeMo/pull/4410#discussion_r904108591),  I guess I never bothered to remove the non necessary non tarred parameters. 

If I understand correctly, all GPUs will still receive an equal amount of data, but the excess tar files (the portion not divisible by the number of used GPUS) will be discarded (i.e. at max `n-1` tar files will be dropped when using `n` GPUs). Bad from the standpoint of using all the data, but all GPUs should still have data to work with, so this should not cause the issue that I am seeing, right?

Unfortunately I can not share the datasets.

This is what I have at the moment:
- the tarred train dataset has 146 files
- the tarred val dataset has 25 files
- the same code ran on 2 GPUs for 55 epochs (starting from scratch) with no issues
- I tried to resume it on 8 GPus, but at the end of epoch 56 there was the `UserWarining` and no validation check, so I started investigating (is it at all possible to resume with a different number of GPUs then initially started?)
- on 4 or 8 GPUs I don't see the end of epoch validation check even if I start from scratch, with `trainer.val_check_interval=0.25` I see the first three validation checks run, but not the one at the end of the epoch
- on 2 GPUs with `trainer.val_check_interval=1.0, trainer.limit_train_batches=0.10, trainer.limit_val_batches=0.10` I can train from scratch as well as resume (though in the resume I see negative numbers for it/s and remaining time, but I assume this is another issue, perhaps related to #4541)
- on 2 GPUs retesting from scratch there is no `UserWarning` and I see the validation check run at the end of the epoch 
- in a couple of hours I'll see what happens if I resume from epoch 55 with 2 GPUs 

At the moment it points out to an issue when the number of GPUs is increased above 2.

**Comment by itzsimpl** [comment]
@PeganovAnton update:
- resuming with 2 GPUs from epoch 55 works, i.e. the Validation check at the end of epoch runs normally
- as soon as I increase the number of GPUs above 2, the Validation check at the end of epoch is simply not run, regardless if I start training from scratch or resume a previous training, and I see the `UserWarning: … could not find the monitored key in the returned metrics`
- on 3 GPUs if I set`trainer.limit_train_batches=0.99` the Validation check at the end of epoch still does not run 
- on 3 GPUs if I set`trainer.limit_train_batches=0.05` or `trainer.limit_train_batches=0.50` the Validation check at the end of epoch runs normally
- on 1 GPU with using the entire tarred train and val sets the Validation check at the end of epoch runs normally
- with a non tarred `validation_ds` on 6 GPUs the Validation check at the end of epoch is not run, hence it a tarred val dataset is not at fault
- with a non tarred `validation_ds` on 2 GPUs the Validation check at the end is run normaly
- since the tarred train set contains 146 files, which nicely splits over 2 GPUs, I assume there is nothing wrong with the train dataset, otherwise training would fail on 2 GPUs as well, this all points to some other issue present

  Indeed observing `lr_scheduler.max_steps` something was off, there is a bug in the computation of `DataLoader.length` for tarred datasets, when these are distributed over multiple gpus and the number of tar files is nondivisble with the number of gpus. I have opened a PR #4553 with a fix.
